### ISSUE TITLE
Change in behaviour of pipe backend protocol 

### ISSUE THREAD

**Issue by cheribral** [PowerDNS/pdns#4308]
When updating a server from v3 to v4 of authoritative PDNS, I've noticed the the behaviour of FAIL in the protocol has changed.  In v3, a FAIL would stop the server from progressing to the next backend, and would return SRVFAIL to a client.  Now, it seems that the backend in simply killed, and the server continues on to the next backend.  

I'm not sure if this was an accidental feature that we happened to use, but the ability for the pipe backend to distinguish between "I don't know, ask someone else" and "The answer definitely doesn't exist, stop looking" was important for us.

Is this a bug in the new version, or the intended behaviour?

**Comment by Habbie** [comment]
To signal that an answer authoritatively does not exist, you should answer for SOA and then send an empty response for the actual question. SERVFAIL is bad form. I'm not sure I agree FAIL should mean 'ignore the next backend'.

I do agree, however, that FAIL should not kill the backend. We've had this issue with the SQL backends during development of 4 but we fixed it there; maybe we missed pipebackend in fixing that.

**Comment by cheribral** [comment]
The problem is really that all the backends are working within the same domains, and they share the same knowledge.  In this case, the pipe backend is picking off query types it needs to handle with more intelligence, and letting the other backend (postgres) do the rest.  Just sending an empty response doesn't work, since postgres will always blithely answer with several matching records, but the pipe backend knows enough about the client to determine if it should get postgres' answer or not.  

Without the hard stop, there is no way for the pipe backend to signal that nothing would work for the client.   I suppose this could be accomplished by writing a much more elaborate backend which does everything postgres and the pipe backend does, but the old accidental behaviour was a pretty handy pipeline for keeping things simple.

**Comment by pieterlexis** [comment]
I think what you were using was a side-effect. The current behaviour is more correct. As @Habbie says: If you want to tell PDNS that the name really doesn't exist, send a SOA when asked for the domain and send no data (just an `END`) when asked for the non-existing name.

This said, the current behaviour is indeed wrong. When the backend-script returns "FAIL", a SERVFAIL should still be sent out and no other backends should be tried. And the inner process should not be restarted.

**Comment by cheribral** [comment]
Ok, thank you.  So to summarize, to do it correctly, every request type needs to be handled entirely in one backend rather than having one backend which does, for example, just manipulating A records while letting everything else (SOA etc) fall through.  The different backends are intended rather for distinct domains.

I was thinking of the backend pipeline perhaps more like a PAM stack for answering requests when it wasn't intended for that.

Basically, yes it was a bug, and I'm doing it wrong :)
